test/performance/odp_ipsec.c

```diff
@@ -654,38 +669,57 @@ run_measure_one(ipsec_args_t *cargs,
 
 	fill_time_record(start);
 
-	while ((packets_sent < cargs->iteration_count) ||
-	       (packets_received < cargs->iteration_count)) {
-		if ((packets_sent < cargs->iteration_count) &&
-		    (packets_sent - packets_received <
-		     cargs->in_flight)) {
-			odp_packet_t out_pkt;
-			int num_out = 1;
+	while ((packets_sent < cargs->packet_count) ||
+	       (packets_received < cargs->packet_count)) {
+		num_enq = cargs->packet_count - packets_sent;
+
+		/* Enqueue up to burst size */
+		num_enq = num_enq > burst_size ? burst_size : num_enq;
+
+		/* Enqueue up to (max in flight - current in flight) */
+		in_flight = packets_sent - packets_received;
+		pkts_allowed = max_in_flight - in_flight;
+		num_enq = num_enq > pkts_allowed ? pkts_allowed : num_enq;
+
+		if (odp_likely(num_enq)) {
+			odp_packet_t out_pkt[num_enq];
+			int i;
 
-			pkt = make_packet(pkt_pool, payload_length);
-			if (ODP_PACKET_INVALID == pkt)
+			if (odp_unlikely(make_packet_multi(pkt_pool,
+							   payload_length,
+							   pkt,
+							   num_enq)))
 				return -1;
 
-			if (cargs->debug_packets)
-				odp_packet_print_data(pkt, 0,
-						      odp_packet_len(pkt));
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_enq; i++)
+					odp_packet_print_data(pkt[i], 0,
+							      odp_packet_len(pkt[i]));
+			}
+
+			num_out = num_enq;
 
-			rc = odp_ipsec_out(&pkt, 1,
-					   &out_pkt, &num_out,
+			rc = odp_ipsec_out(pkt, num_enq,
+					   out_pkt, &num_out,
 					   &param);
-			if (rc <= 0) {
-				ODPH_ERR("failed odp_ipsec_out: rc = %d\n", rc);
-				odp_packet_free(pkt);
+			if (odp_unlikely(rc <= 0)) {
+				ODPH_ERR("Failed odp_ipsec_out: rc = %d\n", rc);
+				odp_packet_free_sp(pkt, num_enq);
 				break;
 			}
-			check_ipsec_result(out_pkt);
+
+			for (i = 0; i < num_out; i++)
+				check_ipsec_result(out_pkt[i]);
 
 			packets_sent += rc;
 			packets_received += num_out;
-			if (cargs->debug_packets)
-				odp_packet_print_data(out_pkt, 0,
-						      odp_packet_len(out_pkt));
-			odp_packet_free(out_pkt);
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_out; i++)
+					odp_packet_print_data(out_pkt[i], 0,
+							      odp_packet_len(out_pkt[i]));
+			}
+
+			odp_packet_free_sp(out_pkt, num_enq);
```

Comment:
This can try to free invalid or stale packet handles. The number of packets in the array is not num_enq but num_out, which may be less than num_enq.

Reply:
If out_pkt is replaced with pkt, that should be enough, right?

Reply:
Yes.

Reply:
I'll make this change as part of v4.

Reply:
Sorry, I misread the question when I answered. Replacing num_enq with out_pkt would have fixed the problem. Replacing out_pkt with pkt is wrong.

odp_ipsec_out() consumes some or all of the input packets and produces output packets. The output packets are not necessarily the same as the input packets although they can be. So after odp_ipsec_out() call you have a zero or more packets left in the input array and you need to free them in order not to leak them. You also have zero or more packets in the output array, but they are new packets and cannot be freed using the old input array handles that are now stale (except that in some implementations, like linux-gen, they may sometimes (but not always), be the same handles in which case the current code accidentally works).

Reply:
Agreed. Will have this changed in v5.